<a href="https://colab.research.google.com/github/xin-2001/taica_1132_GenAI/blob/main/HW11/113_2GenAI_HW11.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Fooocus圖片生成

AI 圖像生成創作任務：打造你的 Fooocus Workflow !

(使用fooocus進行文字生圖。)

作業說明：

1. 設想一個應用情境，並使用 Fooocus 生成圖像。
  > (應用情境可能是：社群平台用圖、簡報圖、網站視覺元素、個人品牌圖像等。)
2. 請至少生成3組圖像，同一張圖的輸入/輸出視為一組。
  > 輸入：為每張圖撰寫簡短說明，例如：使用到 Fooocus 的哪些功能。(prompt 設定、Style、Inpaint、Canny 等）
  > 輸出：生成的圖。
3. 請簡要整理你這份作業的創作流程(文字敘述或流程圖均可)，讓助教/老師能了解你從靈感、設定、試圖改善到輸出圖片的步驟。
4. 整體使用心得分享。
5. 其餘延伸由同學自由發揮。

## 安裝 Fooocus
執行後會產生連結

In [ ]:
!pip install pygit2==1.15.1
%cd /content
!git clone https://github.com/lllyasviel/Fooocus.git
%cd /content/Fooocus
!python entry_with_update.py --share --always-high-vram


   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.1/5.1 MB 54.8 MB/s eta 0:00:00
  Attempting uninstall: pygit2
    Found existing installation: pygit2 1.18.0
    Uninstalling pygit2-1.18.0:
      Successfully uninstalled pygit2-1.18.0
/content
Cloning into 'Fooocus'...
remote: Enumerating objects: 6725, done.
remote: Total 6725 (delta 0), reused 0 (delta 0), pack-reused 6725 (from 1)
Receiving objects: 100% (6725/6725), 33.33 MiB | 25.70 MiB/s, done.
Resolving deltas: 100% (3854/3854), done.
/content/Fooocus
Already up-to-date
Update succeeded.
[System ARGV] ['entry_with_update.py', '--share', '--always-high-vram']
Python 3.11.12 (main, Apr  9 2025, 08:55:54) [GCC 11.4.0]
Fooocus version: 2.5.5
Error checking version for torchsde: No package metadata was found for torchsde
Installing requirements
[Cleanup] Attempting to delete content of temp dir /tmp/fooocus
[Cleanup] Cleanup successful
Downloading: "https://huggingface.co/lllyasviel/misc/resolve/main/xlvaeapp.pth" to /content/Fooocus/mo

## 介面介紹

### 基本介面
![image.png](https://github.com/xin-2001/taica_1132_GenAI/blob/main/HW11/hw11-1.png?raw=true)
1. 主圖像預覽區域 (中央灰色區塊)
  - 這個區域是預覽你即將生成或已經生成的圖像的地方。
  - 在你輸入提示詞並按下「Generate」之前，這裡會顯示一個預設的圖示。
  - 生成圖像後，結果會顯示在這個區域。
2. 文字輸入框 (Prompt)
  - 這是你輸入文字提示詞 (Prompt) 的主要區域。你可以在這裡詳細描述你想要生成的圖像內容、風格、氛圍等等。
  - 輸入框內有預設文字 "Type prompt here or paste parameters."，提示你可以在這裡輸入提示詞，或者如果你有之前儲存的參數，也可以直接貼上。
3. 生成按鈕
  - 當你輸入完提示詞並完成其他設定後，點擊這個 「Generate」 按鈕，Fooocus 就會開始根據你的指示生成圖像。
4. 其他功能設定
  - 還可以額外設定輸入圖像 (Input Image) 選項、Enhance (增強) 選項、Advanced (進階) 選項。

### Input Image

#### Upscale or Variation (放大或變異)
![image.png](https://github.com/xin-2001/taica_1132_GenAI/blob/main/HW11/hw11-2.png?raw=true)
- 左側：圖像輸入區
  > Drop Image Here / Click to Upload： 你可以在這裡拖曳圖片進來，或者點擊上傳按鈕來選擇你想要處理的圖片。
- 右側：Upscale or Variation 設定
  > Upscale or Variation： 這裡讓你選擇對上傳的圖像進行放大 (Upscale) 還是生成變異 (Variation)。

  - Disabled： 停用放大或變異功能。
  - Vary (Subtle)： 生成與原始圖像相似但有細微變化的新圖像。
  - Vary (Strong)： 生成與原始圖像有較大差異的新圖像。
  - Upscale (1.5x)、Upscale (2x)、Upscale (Fast 2x)： 提供不同倍率的圖像放大選項。Fast 2x 通常使用更快速的演算法，可能在細節上略有不同。
- 用途： 這個功能主要用於提升現有圖像的解析度，或者基於現有圖像生成風格相似的新變體。

#### Image Prompt
![image.png](https://github.com/xin-2001/taica_1132_GenAI/blob/main/HW11/hw11-2_2.png?raw=true)
- 這個畫面主要用於讓你上傳多張參考圖像，並設定它們對最終生成圖像的影響程度。
- 有四個相同的圖像上傳區域。每個區域都標示著：Drop Image Here (將圖像拖曳至此)或Click to Upload (點擊上傳)。
  > 功能的主要目的是讓你可以輸入多個不同的參考圖像，並透過後續的設定 (在更進階的設定中) 來調整每一張圖像對最終生成結果的影響力。

  - 風格混合： 你可以上傳不同藝術風格的圖像，讓 AI 學習並融合這些風格到你想要生成的圖像中。
  - 構圖引導： 你可以上傳具有你喜歡的構圖的圖像，引導 AI 在生成新圖像時遵循類似的構圖。
  - 元素參考： 你可以上傳包含特定物體、人物或場景的圖像，讓 AI 在生成新圖像時將這些元素納入考量。
  - 多角度參考： 你可以上傳同一物體或場景的不同角度的照片，幫助 AI 更好地理解其三維結構。
- Advanced:更精細的控制 (例如每張圖像的權重)

#### Inpaint or Outpaint
![image.png](https://github.com/xin-2001/taica_1132_GenAI/blob/main/HW11/hw11-2_3.png?raw=true)
- 這個功能非常強大，它允許你對圖像的特定區域進行修改 (Inpaint) 或擴展 (Outpaint)。
- 圖像編輯預覽區
  - Drop Image Here / Click to Upload： 你需要在這裡拖曳或點擊上傳你想要進行局部修改或擴展的圖像。
  - 遮罩工具 (上傳圖像後顯示)： 一旦你上傳了圖像，這個區域會變成一個圖像編輯器，你會看到一些工具，例如：
    - 筆刷 (Brush)： 用於在圖像上繪製遮罩，選取你想要修改或擴展的區域。通常可以調整筆刷的大小。
    - 橡皮擦 (Eraser)： 用於擦除已經繪製的遮罩。
    - 其他可能的選取工具： 例如矩形選取、套索選取等。
  - 遮罩顯示： 你繪製的遮罩通常會以不同的顏色覆蓋在圖像上，讓你清楚看到你選取的區域。
- Enable Advanced Masking Features (啟用進階遮罩功能)
  > 開啟更精細的遮罩控制功能，例如更複雜的選取模式、羽化邊緣等。在基礎使用中，通常不需要勾選。
- Method (方法)
  - Inpaint or Outpaint (default) (局部重繪或擴展 (預設))：
    > 這是預設選項。當你選擇這個時，Fooocus 會根據你繪製的遮罩以及你輸入的提示詞，自動判斷是應該在遮罩區域內進行局部重繪 (Inpaint) 以替換原有內容，還是應該從遮罩區域向外進行擴展 (Outpaint) 以延伸圖像。
    > Fooocus 會嘗試根據上下文和提示詞的內容來做出最合理的判斷。
  - Improve Detail (face, hand, eyes, etc.) (改善細節 (臉部、手、眼睛等))：
    > 這個選項明確告訴 Fooocus，你希望改善遮罩區域內的現有內容的細節。
    > 它特別提到了「臉部、手、眼睛等」，表明這個選項通常用於提升這些對象的清晰度、紋理和精細度。
    > 當你選擇這個選項時，Fooocus 會嘗試在遮罩區域內保留原有的結構和內容，並在其基礎上增加更多細節。
  - Modify Content (add objects, change background, etc.) (修改內容 (添加物體、改變背景等))：
    > 這個選項明確告訴 Fooocus，你希望修改遮罩區域內的內容。
    > 它提到了「添加物體、改變背景等」作為例子，表明這個選項適用於你想要在遮罩區域內替換掉原有內容，或者添加新的元素。
    > 當你選擇這個選項時，你需要提供清晰的提示詞，描述你希望在遮罩區域生成的新內容。
- Outpaint Direction (擴展方向)
  > 這個區域會顯示出來，讓你指定想要擴展圖像的方向：Left (左)、Right (右)、Top (上)、Bottom (下)。可以選擇一個或多個方向進行擴展。

#### Describe
![image.png](https://github.com/xin-2001/taica_1132_GenAI/blob/main/HW11/hw11-2_4.png?raw=true)


#### Enhance
![image.png](https://github.com/xin-2001/taica_1132_GenAI/blob/main/HW11/hw11-2_5.png?raw=true)


#### Metadata
![image.png](https://github.com/xin-2001/taica_1132_GenAI/blob/main/HW11/hw11-2_6.png?raw=true)



### Enhance


#### Upscale or Variation (放大或變異)
![image.png](https://github.com/xin-2001/taica_1132_GenAI/blob/main/HW11/hw11-3.png?raw=true)


#### 更進階的增強設定區域(#1、#2、#3)
![image.png](https://github.com/xin-2001/taica_1132_GenAI/blob/main/HW11/hw11-3_2.png?raw=true)

### Advanced


#### Settings (設定)
![image.png](https://github.com/xin-2001/taica_1132_GenAI/blob/main/HW11/hw11-4.png?raw=true)

#### Styles (風格)
![image.png](https://github.com/xin-2001/taica_1132_GenAI/blob/main/HW11/hw11-4_2.png?raw=true)

#### Models (模型)
![image.png](https://github.com/xin-2001/taica_1132_GenAI/blob/main/HW11/hw11-4_3.png?raw=true)

#### Advanced (進階)
![image.png](https://github.com/xin-2001/taica_1132_GenAI/blob/main/HW11/hw11-4_4.png?raw=true)



## 結果展現
- Prompt:A shiba inu slumped over a desk full of computers and coffee cups, exhausted but cute.
- Enhance:Illustration in soft pastel tones, gentle hand-drawn look with watercolor or thick painterly strokes. Minimalistic and clean background, cozy and cute atmosphere with a hint of playful mischief. Characters are outlined with slightly thick lines. Flat coloring with subtle texture. A small Shiba Inu dog subtly hidden in the background, not stealing focus. Whimsical, heartwarming, and adorable overall style.
- Negative:ugly animal,unattractive pet,poorly groomed fur,messy coat,awkward pose,bad anatomy,mutated limbs,deformed body,low quality image,blurry photo,jpeg artifacts,bad proportions,missing limbs,extra limbs,cropped image,worst quality,low resolution,watermark,signature,distorted face, unnatural expression,exaggerated ssmile,weirdeyes,awkward facial pose,unappealing face,strange mouth,bad facial anatomy,mutated face,deformed features, creepy expression

1.1 僅prompt
  > ![image.png](https://github.com/xin-2001/taica_1132_GenAI/blob/main/HW11/result/result-1.png?raw=true)



2.1 勾選Input image>Upscale or Variation，並選擇Vary (Strong)
  > ![image.png](https://github.com/xin-2001/taica_1132_GenAI/blob/main/HW11/result/result-2.png?raw=true)

2.2 勾選Input image>Upscale or Variation，並選擇Vary (Subtle)
  > ![image.png](https://github.com/xin-2001/taica_1132_GenAI/blob/main/HW11/result/result-3.png?raw=true)

2.3 勾選Input image>Image Prompt(無勾選Advanced)
  > ![image.png](https://github.com/xin-2001/taica_1132_GenAI/blob/main/HW11/result/result-4.png?raw=true)

2.4 勾選Input image>Image Prompt(勾選Advanced)
  - 將兩張圖片Weight設置為0.8
  > ![image.png](https://github.com/xin-2001/taica_1132_GenAI/blob/main/HW11/result/result-5.png?raw=true)

2.5 勾選Input image>Image Prompt(勾選Advanced)
  - 將左圖片Weight設置為1，右圖Weight設置為1.2
  > ![image.png](https://github.com/xin-2001/taica_1132_GenAI/blob/main/HW11/result/result-6.png?raw=true)

2.6 勾選Input image>Describe
  - 經過圖片生成的prompt=```closed mouth, closed eyes, cup, no humans, animal, phone, sleeping, cellphone, smartphone, desk, dog, mug, pen, computer, disposable cup, pencil, animal focus, monitor, coffee, keyboard \(computer\), mouse \(computer\), coffee cup, eraser, shiba inu```
  > ![image.png](https://github.com/xin-2001/taica_1132_GenAI/blob/main/HW11/result/result-7.png?raw=true)

3.1 勾選Advanced>setting
  - Performance選擇Quality
  - 設置Negative Prompt
  - 設置Aspect Ratios為1:1
  > ![image.png](https://github.com/xin-2001/taica_1132_GenAI/blob/main/HW11/result/result-8.png?raw=true)